File: dec24/gosolutions/day10/10.go
package day10

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"runtime"
)

type Dir struct {
	dx, dy int
}

var (
	rows, cols int
	north = Dir{dx: -1, dy: 0}
	east  = Dir{dx: 0, dy: 1}
	south = Dir{dx: 1, dy: 0}
	west  = Dir{dx: 0, dy: -1}

	dirs = []Dir{north, east, south, west}
)

type Pos struct {
	x, y int
}

func (pos *Pos) valid() bool {
	return pos.x >= 0 && pos.x < rows && pos.y >= 0 && pos.y < cols;
}

func (pos *Pos) move(dir Dir) Pos {
	return Pos{x: pos.x + dir.dx, y: pos.y + dir.dy}
}

func dfs(input []string, pos Pos, currScore map[Pos]int, ) {
	if !pos.valid(){
		return
	}

	if input[pos.x][pos.y] == '9' {
		currScore[pos] += 1
		return
	}

	for _, dir := range dirs {
		newPos := pos.move(dir)
		if newPos.valid() && input[newPos.x][newPos.y] == input[pos.x][pos.y] + 1 {
			dfs(input, newPos, currScore)
		}
	}
}

// part a
func solve1(input []string) int {
	score := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if input[i][j] != '0' {
				continue
			}
			currScore := make(map[Pos]int, 0)
			dfs(input, Pos{x: i, y: j}, currScore)
			score += len(currScore)
		}
	}
	return score
}

// part b
func solve2(input []string) int {
	score := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if input[i][j] != '0' {
				continue
			}
			currScore := make(map[Pos]int, 0)
			dfs(input, Pos{x: i, y: j}, currScore)

			// add all scores to total score
			for _, val := range currScore {
				score += val
			}
		}
	}
	return score
}

func Day10() {
	// Get the path of the current Go source file
	_, currentFile, _, _ := runtime.Caller(0)

	// Get the directory of the Go source file
	currentDir := filepath.Dir(currentFile)
    parentDir := filepath.Dir(filepath.Dir(currentDir))
    file, err := os.Open(parentDir + "/testcases/10.txt")

    if err != nil {
        log.Fatal("Unable to open testcase file")
        return
    }

    defer file.Close()

    scanner := bufio.NewScanner(file)
	input := make([]string, 0)
	for scanner.Scan() {
		input = append(input, scanner.Text())
	}

	rows = len(input)
    cols = len(input[0])

	// compute answers
	fmt.Println("Expected answers:")
    fmt.Println("Part a: ", 36)
    fmt.Println("Part b: ", 81)
    fmt.Println("Your answers:")
    fmt.Println("Part a: ", solve1(input))
    fmt.Println("Part b: ", solve2(input))
}

File: cppsolutions/10.cpp
#include<iostream>
#include <fstream>
#include<vector>
#include <string>
#include <sstream>
#include <unordered_set>
using namespace std;

bool valid(vector<string> &input, int i, int j){
    return i >= 0 && i < input.size() && j >= 0 && j < input[0].size();
}

string getKey(int i, int j){
    return to_string(i) + "-" + to_string(j);
}

vector<vector<int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

void dfs(vector<string> &input, int i, int j, unordered_set<string> &score){
    if (!valid(input, i, j)){
        return;
    }
    // cout << "i: " << i << ", j: " << j << ", val:" << input[i][j] << endl;
    if(input[i][j] == '9'){
        score.insert(getKey(i, j));
        return;
    }

    for(vector<int> &dir : dirs){
        int nr = i + dir[0], nc = j + dir[1];

        if (valid(input, nr, nc) && input[nr][nc] == input[i][j] + 1 ){
            dfs(input, nr, nc, score);
        }
    }
}

int solve1(vector<string> &input){
    int totalScore = 0;
    for(int i = 0; i < input.size(); i++){
        for(int j = 0; j < input[0].size(); j++){
            if(input[i][j] == '0'){
                unordered_set<string> score;
                dfs(input, i, j, score);
                totalScore += score.size();
                // cout << "score: " << i << ","<< j << " = " << score.size() << endl;
            }
        }
    }
    return totalScore;
}

void dfs2(vector<string> &input, int i, int j, int &score){
    if (!valid(input, i, j)){
        return;
    }
    // cout << "i: " << i << ", j: " << j << ", val:" << input[i][j] << endl;
    if(input[i][j] == '9'){
        score++;
        return;
    }

    for(vector<int> &dir : dirs){
        int nr = i + dir[0], nc = j + dir[1];

        if (valid(input, nr, nc) && input[nr][nc] == input[i][j] + 1){
            dfs2(input, nr, nc, score);
        }
    }
}


int solve2(vector<string> &input){
    int totalScore = 0;
    for(int i = 0; i < input.size(); i++){
        for(int j = 0; j < input[0].size(); j++){
            if(input[i][j] == '0'){
                dfs2(input, i, j, totalScore);
                // cout << "score: " << i << ","<< j << " = " << score.size() << endl;
            }
        }
    }
    return totalScore;
}

int main() {
    ifstream f("../testcases/10.txt");

    if (!f) {
        cerr << "Error opening the file!" << endl;
        return 1;
    }

    string s;
    vector<string> input;
    while(getline(f, s)){
        input.push_back(s);
    }

    // part a
    long long partA = solve1(input);
    cout << "Part a: " << partA << endl;

    // part b
    long long partB = solve2(input);
    cout << "Part b: " << partB << endl;
    
    // Expected answers:
    // Part a: 557
    // Part b: 1062
    f.close();
    
    return 0;
}